lib/galaxy/model/__init__.py

```diff
@@ -657,6 +657,35 @@ def calculate_user_disk_usage_statements(user_id, quota_source_map, for_sqlite=F
     return statements
 
 
+UNIQUE_DATASET_USER_USAGE_PER_OBJECTSTORE = """
+WITH per_user_histories AS
+(
+    SELECT id
+    FROM history
+    WHERE user_id = :id
+        AND NOT purged
+),
+per_hist_hdas AS (
+    SELECT DISTINCT dataset_id
+    FROM history_dataset_association
+    WHERE NOT purged
+        AND history_id IN (SELECT id FROM per_user_histories)
+)
+SELECT COALESCE(SUM(COALESCE(dataset.total_size, dataset.file_size, 0)), 0), dataset.object_store_id
```

Comment:
Why do we need 2 coalesce calls here? As I understand, the inner coalesce guarantees the value supplied to sum is already never null?

Reply:
Dropped it - we will see if any of the tests break.

Reply:
TestCalculateUsage.test_dictify_usage_unused_quota_labels breaks when I dropped the same "extra" coalesce ``UNIQUE_DATASET_USER_USAGE`` ... that is used as part of an insert and resulted in a null being added to a NON NULL number column:

```
E       sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) NOT NULL constraint failed: user_quota_source_usage.disk_usage
E       [SQL: 
E       WITH new (user_id, quota_source_label, disk_usage) AS (
E           VALUES(?, ?, (
E       WITH per_user_histories AS
E       (
E           SELECT id
E           FROM history
E           WHERE user_id = ?
E               AND NOT purged
E       ),
E       per_hist_hdas AS (
E           SELECT DISTINCT dataset_id
E           FROM history_dataset_association
E           WHERE NOT purged
E               AND history_id IN (SELECT id FROM per_user_histories)
E       )
E       SELECT SUM(COALESCE(dataset.total_size, dataset.file_size, 0))
E       FROM dataset
E       LEFT OUTER JOIN library_dataset_dataset_association ON dataset.id = library_dataset_dataset_association.dataset_id
E       WHERE dataset.id IN (SELECT dataset_id FROM per_hist_hdas)
E           AND library_dataset_dataset_association.id IS NULL
E           AND ( dataset.object_store_id IN (?) )
E       ))
E       )
E       INSERT OR REPLACE INTO user_quota_source_usage (id, user_id, quota_source_label, disk_usage)
E       SELECT old.id, new.user_id, new.quota_source_label, new.disk_usage
E       FROM new
E           LEFT JOIN user_quota_source_usage AS old
E               ON new.user_id = old.user_id
E                   AND new.quota_source_label = old.quota_source_label
E       ]
E       [parameters: (7, 'unused_source', 7, 'unused_source_store')]
E       (Background on this error at: https://sqlalche.me/e/14/gkpj)
```

I will keep your requested change because I think in this place it is only used with selects and not joined against a table that might result in no datasets being selected and the SUM product being null.